src/modal/modal-window.ts

```diff
@@ -52,40 +55,56 @@ export class NgbModalWindow implements OnInit,
   @Output('dismiss') dismissEvent = new EventEmitter();
 
   constructor(
-      @Inject(DOCUMENT) private _document: any, private _elRef: ElementRef<HTMLElement>, private _zone: NgZone) {
-    _zone.runOutsideAngular(() => {
-      fromEvent<KeyboardEvent>(this._elRef.nativeElement, 'keydown')
+      @Inject(DOCUMENT) private _document: any, private _elRef: ElementRef<HTMLElement>, private _zone: NgZone) {}
+
+  dismiss(reason): void { this.dismissEvent.emit(reason); }
+
+  ngOnInit() { this._elWithFocus = this._document.activeElement; }
+
+  ngAfterViewInit() {
+    const {nativeElement} = this._elRef;
+    this._zone.runOutsideAngular(() => {
+
+      fromEvent<KeyboardEvent>(nativeElement, 'keydown')
           .pipe(
               takeUntil(this.dismissEvent),
               // tslint:disable-next-line:deprecation
               filter(e => e.which === Key.Escape && this.keyboard))
           .subscribe(event => requestAnimationFrame(() => {
                        if (!event.defaultPrevented) {
-                         _zone.run(() => this.dismiss(ModalDismissReasons.ESC));
+                         this._zone.run(() => this.dismiss(ModalDismissReasons.ESC));
                        }
                      }));
 
-      const mouseDowns$ = fromEvent<MouseEvent>(this._elRef.nativeElement, 'mousedown')
-                              .pipe(
-                                  takeUntil(this.dismissEvent),
-                                  map(e => this.backdrop === true && this._elRef.nativeElement === e.target));
+      // We're listening to 'mousedown' and 'mouseup' to prevent modal from closing when pressing the mouse
+      // inside the modal dialog and releasing it outside
+      let preventClose = false;
+      fromEvent<MouseEvent>(this._dialogEl.nativeElement, 'mousedown')
+          .pipe(takeUntil(this.dismissEvent), tap(() => {
+                  preventClose = false;
+                  fromEvent<MouseEvent>(nativeElement, 'mouseup')
+                      .pipe(takeUntil(this.dismissEvent), take(1))
+                      .subscribe(({target}) => { preventClose = nativeElement === target; });
+                }))
+          .subscribe();
```

Comment:
I think we could safely rewrite this part as

```typescript
.pipe(
  takeUntil(this.dismissEvent),
  tap(() => { preventClose = false; }),
  switchMap(() => fromEvent<MouseEvent>(nativeElement, 'mouseup')
    .pipe(takeUntil(this.dismissEvent), take(1))
  )
).subscribe(({target}) => { preventClose = nativeElement === target; });
```

Reply:
Done. Thanks, this looks much better!

```ts
      fromEvent<MouseEvent>(this._dialogEl.nativeElement, 'mousedown')
          .pipe(
              takeUntil(this.dismissEvent), 
              tap(() => preventClose = false),
              switchMap(() => fromEvent<MouseEvent>(nativeElement, 'mouseup').pipe(takeUntil(this.dismissEvent), take(1))),
              filter(({target}) => nativeElement === target))
          .subscribe(() => { preventClose = true; });
```